fenicsprecice/adapter_core.py

```diff
@@ -434,27 +437,54 @@ def get_forces_as_point_sources(fixed_boundary, function_space, data):
     """
     x_forces = dict()  # dict of PointSources for Forces in x direction
     y_forces = dict()  # dict of PointSources for Forces in y direction
+    z_forces = dict()  # dict of PointSources for Forces in z direction
 
-    fenics_vertices = np.array(list(data.keys()))
+    vertices = np.array(list(data.keys()))
     nodal_data = np.array(list(data.values()))
 
     # Check for shape of coupling_mesh_vertices and raise Assertion for 3D
-    n_vertices, _ = fenics_vertices.shape
-
-    vertices_x = fenics_vertices[:, 0]
-    vertices_y = fenics_vertices[:, 1]
-
-    for i in range(n_vertices):
-        px, py = vertices_x[i], vertices_y[i]
-        key = (px, py)
-        x_forces[key] = PointSource(function_space.sub(0), Point(px, py), nodal_data[i, 0])
-        y_forces[key] = PointSource(function_space.sub(1), Point(px, py), nodal_data[i, 1])
-
-    # Avoid application of PointSource and Dirichlet boundary condition at the same point by filtering
-    x_forces = filter_point_sources(x_forces, fixed_boundary, warn_duplicate=False)
-    y_forces = filter_point_sources(y_forces, fixed_boundary, warn_duplicate=False)
-
-    return x_forces.values(), y_forces.values()  # don't return dictionary, but list of PointSources
+    n_vertices, _ = vertices.shape
+
+    vertices_x = None
+    vertices_y = None
+    vertices_z = None
```

Comment:
Is this needed? Could we get rid of the `vertices_x`, `vertices_y`, `vertices_z` and use something more generic? This might help us to put the whole code into a loop of the following form:
```
for d in range(dims):
    for i in range(n_vertices):
        p = get_point(vertices, i)  # not sure how to solve this, but should be possible
        key = get_key(p)
        forces[d, key] = PointSource(function_space.sub(d), p, nodal_data[i, d])
    
    forces[d, :] = filter_point_sources(forces[d, :], fixed_boundary, warn_duplicate=False)
return make_tuple(forces)  # to make sure you do not break the interface
```

Again: be caseful. Pseudocode, but you should get the idea.

Reply:
This is possible but not really related to 3D functionality. A generic solution can be added through this PR though, as handling different dimensions separately does make it ugly.